lib/fs.sh

```diff
@@ -23,7 +23,8 @@ function home_exists {
 function list_castle_names {
 	while IFS= read -d $'\0' -r repo ; do
 		local reponame
-		reponame=$(basename "${repo%/.git}")
+		reponame="${repo%/.git}"
+		reponame="${reponame##*/}"
```

Comment:
[See here](https://stackoverflow.com/a/22402242/339505) why I'm using `basename`, there are quite a few pitfalls. Doing it the proper way, we should probably even replace `%/.git` with `dirname`. It also increases readability.  
Even if we could exclude all the cases where `##*/` would fail, it's just not worth it from a maintainability and stability standpoint.

Reply:
Okay, there's only `k` calls to `basename` during a no-op refresh, but `4k` calls to `stat`. Maybe caching those results will make up for calling `basename`.

Reply:
Actually, looking at that post it seems the pitfalls don't apply here. You're controlling the input to `basename` with the `find` command below. It's guaranteed to be missing the `/.git` part at the end due to the line directly above this one so it will have a format like `/usr/local/home/myhome/.homesick/repos/foo` at this point.

Reply:
Hm, you're right. There really is no way for this to fail. I would be okay with including this, *if* you add a small comment explaining why we don't need basename. Otherwise it'll be an unnecessary point to think about when debugging some problem with paths because that reasoning is not readily available.

Reply:
You could also get rid of the `find | sort` by just letting `bash` glob into an array. E.g.,
```bash
function list_castle_names {
    local castle_git_dirs=("$repos"/*/.git)
    for d in "${castle_git_dirs[@]}"; do
        local repodir="${d%/.git}"
        printf "%s\n" "${repodir##*/}"
    done
    return "$EX_SUCCESS"
}
```
That should follow symlinks as well as sort the names, but it doesn't check that `.git` is a directory. You could add an extra dir check before printing if you want, but technically `.git` could be a file (e.g., if it's under a Git worktree dir). Also, if there are no matches at all, you would need to set `nullglob` (https://www.gnu.org/software/bash/manual/html_node/Filename-Expansion.html), but presumably, there should always be a `homeshick` castle at least.

Reply:
@jjlin, what about special characters in the castle names? I'm not sure bash globbing would be able to handle things like newlines.

Reply:
I did a quick test and newlines seem to work, though I'd think that would be a pretty unusual thing to do.

Reply:
@andsens I'm saying I don't think it works in `homeshick list`. Since `list_castle_names` would output two castles `castle_no_newline` and `castle\nwith_newline` as

```
castle_no_newline
castle
with_newline
```

when https://github.com/andsens/homeshick/blob/master/lib/commands/list.sh#L4 reads this in, it would interpret it as three castle names.